### PR TITLE
[MM-26997] Use socket API to deploy plugins, if available

### PR DESCRIPTION
**Pull request by mickmister**
#### Summary

This PR makes it so if a local Mattermost server is listening on a unix socket, the plugin deploy scripts will use the local mode API instead of authenticating as a user. This makes it much easier to set up a development environment for working with plugins.

#### Ticket Link

Fixes https://github.com/mattermost/mattermost-plugin-starter-template/issues/107
https://mattermost.atlassian.net/jira/software/c/projects/MM/issues/MM-26997

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -36,7 +36,27 @@ dist/com.example.my-plugin.tar.gz
 
 ## Development
 
-To avoid having to manually install your plugin, build and deploy your plugin with login credentials:
+To avoid having to manually install your plugin, build and deploy your plugin using one of the following options.
+
+### Deploying with Local Mode
+
+If your Mattermost server is running locally, you can configure the server to listen on a unix socket. The plugin deploy scripts can then interact directly with the server without requiring authentication credentials. In order to do this, you need to edit the server's `ServiceSettings` in `mattermost-server/config/config.json` to support [local mode](https://docs.mattermost.com/administration/mmctl-cli-tool.html#local-mode).
```
Comment (lieut-data) [inline README.md:43]:
```suggestion
 If your Mattermost server is running locally, you can enable [local mode](https://docs.mattermost.com/administration/mmctl-cli-tool.html#local-mode) to streamline deploying your plugin. Edit your server configuration as follows:
```

File: README.md [hunk 2]
```diff
@@ -36,7 +36,27 @@ dist/com.example.my-plugin.tar.gz
...
+{
+    "ServiceSettings": {
+        ...
+        "EnableLocalMode": true,
+        "LocalModeSocketLocation": "/var/tmp/mattermost_local.socket"
+    }
+}
+```
+
+If you'd like to use a different path than `/var/tmp/mattermost_local.socket`, assign the intended path to the environment variable `MM_LOCALSOCKETPATH` in your terminal being used to deploy the plugin.
```
Comment (lieut-data) [inline README.md:55]: (can't use the suggestion trick here, but):

---------

and then deploy your plugin:
```
make deploy
```

You may also customize the Unix socket path:
```
export MM_LOCALSOCKETPATH=/var/tmp/alternate_local.socket
make deploy
```

File: README.md [hunk 3]
```diff
@@ -36,7 +36,27 @@ dist/com.example.my-plugin.tar.gz
...
+
```
Comment (lieut-data) [inline README.md:56]: Can't comment on the line way below here, but can we move the `make watch` documentation  up here as well? I think local socket mode is likely going to become "the default", so don't want to bury the `make watch` unnecessarily.

File: build/pluginctl/main.go [hunk 4]
```diff
@@ -85,6 +99,23 @@ func getClient() (*model.Client4, error) {
 	return nil, errors.New("one of MM_ADMIN_TOKEN or MM_ADMIN_USERNAME/MM_ADMIN_PASSWORD must be defined")
 }
 
+func getUnixClient(socketPath string) (*model.Client4, error) {
```
Comment (lieut-data) [inline build/pluginctl/main.go:102]: We're using `error` as a proxy for "found socket" here, but we never actually /check/ the value of the returned error. Should we test the error values to distinguish "not running" vs. "tried but failed for other reasons"? Or should we just return `*model.Client4, bool` to make this more explicit?

File: build/pluginctl/main.go [hunk 5]
```diff
@@ -55,6 +59,16 @@ func pluginctl() error {
...
 
 func getClient() (*model.Client4, error) {
+	socketPath := os.Getenv("MM_LOCALSOCKETPATH")
+	if socketPath == "" {
+		socketPath = defaultSocketPath
+	}
+
+	client, err := getUnixClient(socketPath)
+	if err == nil {
+		return client, nil
```
Comment (lieut-data) [inline build/pluginctl/main.go:69]: Can we log an info message?
```suggestion
 		log.Printf("Connecting using local mode over %s", socketPath)
 		return client, nil  		
```

File: build/pluginctl/main.go [hunk 6]
```diff
@@ -85,6 +99,23 @@ func getClient() (*model.Client4, error) {
...
+
+	tr := &http.Transport{
+		Dial: func(network, addr string) (net.Conn, error) {
+			return net.Dial("unix", socketPath)
+		},
+	}
+	client := model.NewAPIv4Client("http://_")
+	client.HttpClient = &http.Client{Transport: tr}
+
+	return client, nil
```
Comment (mgdelacroix) [inline build/pluginctl/main.go:116]: We can use [the recently merged `NewAPIv4SocketClient` function here](https://github.com/mattermost/mattermost-server/blob/b372b98aca865b037cadcc9b3c53f19f7ed63dc3/model/client4.go#L122):

```suggestion
	return model.NewAPIv4SocketClient(socketPath), nil
```

File: build/pluginctl/main.go [hunk 7]
```diff
@@ -55,6 +59,16 @@ func pluginctl() error {
 }
 
 func getClient() (*model.Client4, error) {
+	socketPath := os.Getenv("MM_LOCALSOCKETPATH")
+	if socketPath == "" {
+		socketPath = defaultSocketPath
```
[the remark this answers is not in the document]
Comment (mickmister) [reply build/pluginctl/main.go:64]: @mgdelacroix I'll have to upgrade the plugin's dependency on mattermost-server for this one situation. We generally try to keep the plugins as low-version compatible as possible.

File: build/pluginctl/main.go (hunk 6 above)
Comment (mickmister) [reply build/pluginctl/main.go]: @mgdelacroix Same comment about this change. I'd like to get @lieut-data's opinion on this.

File: build/pluginctl/main.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (lieut-data) [reply build/pluginctl/main.go:64]: @mickmister, the `go.mod` dependency update here is independent from the `min_server_version`. You should be able to upgrade to the latest `v5.26` or even `master` while keeping the min server version at the last ESR.

[the remark this answers is not in the document]
Comment (mickmister) [reply build/pluginctl/main.go:64]: @lieut-data I was originally concerned about how we will be subsequently propagating this dependency change to all plugins, but that is fine as you have pointed out.

File: build/pluginctl/main.go [hunk 8]
```diff
@@ -70,6 +67,8 @@ func getClient() (*model.Client4, error) {
 		return client, nil
 	}
 
+	log.Printf("No socket found at %s for local mode deployment. Attempting to authenticate with credentials.", socketPath)
```
Comment (lieut-data) [inline build/pluginctl/main.go:70]: nit: this is entirely redundant if I intentionally set the other environment variables. Can we only emit this if `MM_LOCALSOCKETPATH` was non-empty?

Comment (mickmister) [reply build/pluginctl/main.go]: Done!

**Comment by mickmister** [comment]
@mgdelacroix Ping on this. This is (I think 🙂) the last major change before I sync this repo with all of the other plugin repos.

**Pull request merged**